### PR TITLE
ci: release script

### PR DESCRIPTION
**Pull request by kazuk**
## Issue number and link

Relates: #66

## Describe your changes

- add bash script for edit CHANGELOG
- add bumping crate versions
- publish crates
- git tagging and push

currently disable for any update to git and crates.io , so we need test for flow is collect.

## Checklist before requesting a review

- [x] I follow the [Semantic Pull Requests](https://github.com/zeke/semantic-pull-requests) rules (bugfix/feature)
- [x] I specified links to related issues (must: bugfix, want: feature)
- [x] I have performed a self-review of my code (bugfix/feature)
- [ ] I have added thorough tests (bugfix/feature)
- [ ] I have edited `## [Unreleased]` section in `CHANGELOG.md` following [keep a changelog](https://keepachangelog.com/en/1.0.0/) syntax (bugfix/feature)
- [ ] I {made/will make} a related pull request for [documentation repo](https://github.com/SpringQL/SpringQL.github.io) (feature)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+name: release
+
+on:
+  workflow_dispatch:
+    inputs: 
+      current_version:
```
Comment (laysakura) [inline .github/workflows/release.yml:6]: Want to get the current version from somewhere (Cargo.toml, CHANGELOG.md, etc).

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        description: 'current version number'
+        required: true
+        type: string
+      release_version:
+        description: 'release version number'
```
Comment (laysakura) [inline .github/workflows/release.yml:11]:
```suggestion
        description: 'release version (e.g. `v0.15.1`)'
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+          source .github/workflows/scripts/keep-a-changelog.bash
+          # edit CHANGELOG.md
+          cp CHANGELOG.md CHANGELOG.md.org
+          cat CHANGELOG.md.org | bump_changelog "${CURRENT_VERSION}" "${RELEASE_VERSION}" > CHANGELOG.md
+          rm CHANGELOG.md.org
+      
+      - name: bump crate versions
+        run: |
+          cd springql
+          cargo bump "${RELEASE_VERSION}"
```
Comment (laysakura) [inline .github/workflows/release.yml:43]: Does `cargo-bump` update dependency versions?

```toml
# springql/Cargo.toml
# ...
[dependencies]
springql-core = "0.14.0"
```

```bash
cd springql-core
cargo bump 0.15.0
```

Then will the "= 0.14.0" be changed into "= 0.15.0"?

Comment (laysakura) [reply .github/workflows/release.yml:43]: `${RELEASE_VERSION}` contains `v` prefix. Is that OK?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        run: |
+          # TODO: de-comment for login
+          # cargo login "${{ secrets.CRATES_IO_TOKEN }}"
+          cargo publish -p springql-core --dry-run # TODO: remove `--dry-run`
+          cargo publish -p springql --dry-run      # TODO: remove `--dry-run`
+
+      - name: commit changes
+        run: |
+          git add .
+          git commit -m "update changelog for release ${RELEASE_VERSION}"
```
Comment (laysakura) [inline .github/workflows/release.yml:65]: I guess not only the changelog but also `Cargo.toml`s are updated.

File: .github/workflows/scripts/keep-a-changelog.bash [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        if [ "$line" == "[Unreleased]: https://github.com/SpringQL/SpringQL/compare/${CURRENT_VERSION}...HEAD" ]; then 
+            # update line for [Unreleased]
+            line="[Unreleased]: https://github.com/SpringQL/SpringQL/compare/${NEW_VERSION}...HEAD"
+        fi
+
+        echo "$line"
+        if [ "$line" == "## [Unreleased]" ]; then
+            # insert line after ## [Unleleased]
+            echo # output blank line
+            echo "## [${NEW_VERSION}]"
```
Comment (laysakura) [inline .github/workflows/scripts/keep-a-changelog.bash:22]: From v0.14.0, we started to add the release date here.

https://github.com/SpringQL/SpringQL/blob/main/CHANGELOG.md#v0140---2022-06-24

Could you add the date automatically from system clock? (either UTC or JST would be OK)

**Review by laysakura — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 3 above)
Comment (kazuk) [reply .github/workflows/release.yml:43]: cargo-bump is here. https://crates.io/crates/cargo-bump

Comment (laysakura) [reply .github/workflows/release.yml:43]: Yeah, I read the document but I could not figure out these questions from it.

Comment (kazuk) [reply .github/workflows/release.yml:43]: `${RELEASE_VERSION}' and `${CURRENT_VERSION}` must not contains `v`.

`cargo-bump` updates crate package version only. dependencies not changed.
dependency update needs `cargo upgrade` https://crates.io/crates/cargo-edit#cargo-upgrade

File: .github/workflows/release.yml (hunk 1 above)
Comment (kazuk) [reply .github/workflows/release.yml:6]: I think need to know the current release number for the release work.

The version number in `Cargo.toml` may change to a snapshot version and may not be a good value to get and use automatically.

It should be possible to assume that the release worker knows the current version number, even if unexpected release work occurs, such as when re-releasing a fix for a previous version due to a security fix.

Therefore, I think it's better to let the release worker enter the current version it knows.

**Review by laysakura — DISMISSED** [review]
Just a minor comment. You can merge this to try the workflow execution if you will do something for the comment later.

**Comment by laysakura** [comment]
`ci / lcov` job is already fixed in the main branch.

File: .github/workflows/scripts/keep-a-changelog.bash (hunk 5 above)
Comment (laysakura) [reply .github/workflows/scripts/keep-a-changelog.bash:22]: Ah, sorry. How about this?

**Review by laysakura — APPROVED** [review]
You can merge it for checking the workflow but I added a comment.

File: .github/workflows/scripts/keep-a-changelog.bash (hunk 5 above)
Comment (kazuk) [reply .github/workflows/scripts/keep-a-changelog.bash:22]: I can do it with minor change. 
I want tests whole release workflow first.

**Pull request merged**